### PR TITLE
Password reset feature #508

### PR DESCRIPTION
**Pull request by atomfrede**
A proposal for the long wanted password reset feature :)

* Tested with sql/mongo on java 8
* Java 7 should work too
* Not sure about the cassandra version, but @jdubois you are the expert and should have a close look at it

The rest works as follows (see #508 for details):

* User requests a password reset by proving his/her email
* If mail exists a reset key and date are created and a mail is send to the user with reset link
* Link is valid for 24 hours
* Visiting the link shows a password reset form (like password change form)
* After formular is posted the user has the new password (reset key and date are removed)

### CODE REVIEW

**Comment by PeterEltgroth** [comment]
Yay!  Thanks for the work. I look forward to when it is merged into a release.

**Comment by jdubois** [comment]
That's a huge work, thanks!!!
I might use this at Devoxx Fr this week, as we have a half day harckergarten: that could be a good exercise for everyone.
Give me some time to review everything, but indeed this should be merged!!

**Comment by atomfrede** [comment]
Take your time, better to have a great implementation than having to create a lot patches afterwards.

**Comment by gmarziou** [comment]
Great feature and great work!
I'm just wondering whether there should be some mechanism to prevent too many reset requests per user.
Some malicious people could use it to spam another user.

**Comment by atomfrede** [comment]
@gmarziou Sounds like a great idea. Will think about how to implement it.

**Comment by jdubois** [comment]
@atomfrede you forgot to add the email reset template, so when the email is sent we have an error, can you add that file?
Otherwise I started merging your code in https://github.com/jhipster/generator-jhipster/tree/atomfrede-password-reset-feature -> can you have a look at it?

**Comment by atomfrede** [comment]
@jdubois Oh sorry :( Will add it today.

File: app/templates/src/main/webapp/scripts/app/account/reset/finish/reset.finish.html [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                        <div ng-show="form.confirmPassword.$dirty && form.confirmPassword.$invalid">
+                            <p class="help-block"
+                               ng-show="form.confirmPassword.$error.required" translate="global.messages.validate.confirmpassword.required">
+                                Your password confirmation is required.
+                            </p>
+                            <p class="help-block"
+                               ng-show="form.confirmPassword.$error.minlength" translate="global.messages.validate.confirmpassword.minlength">
+                                Your password confirmation is required to be at least 5 characters
+                            </p>
+                            <p class="help-block"
```
Comment (gmarziou) [inline app/templates/src/main/webapp/scripts/app/account/reset/finish/reset.finish.html]: You could simplify markup for confirmation password by removing min and max length as this is already covered  for password and matching both.

**Comment by atomfrede** [comment]
@gmarziou Guess you're right, just copied the registration form in fact

File: app/templates/src/main/webapp/scripts/components/auth/services/_password.service.js [hunk 2]
```diff
@@ -5,3 +5,15 @@ angular.module('<%=angularAppName%>')
         return $resource('api/account/change_password', {}, {
         });
     });
+
+angular.module('jhipsterApp')
```
Comment (gmarziou) [inline app/templates/src/main/webapp/scripts/components/auth/services/_password.service.js]: jhipsterApp should be <%=angularAppName%>

Maybe you should consider just chaining the 3 calls to factory().

File: app/templates/src/main/java/package/web/rest/_AccountResource.java [hunk 3]
```diff
@@ -257,4 +257,54 @@ public void invalidateSession(@PathVariable String series) throws UnsupportedEnc
...
+        <% if (javaVersion == '8') { %>
+        return userService.requestPasswordReset(mail)
+            .map(user -> {
+                String baseUrl = request.getScheme() +
+                    "://" +
+                    request.getServerName() +
+                    ":" +
+                    request.getServerPort();
+            mailService.sendPasswordResetMail(user, baseUrl);
+            return new ResponseEntity<>("e-mail was send", HttpStatus.OK);
```
Comment (gmarziou) [inline app/templates/src/main/java/package/web/rest/_AccountResource.java]: typo: should be "e-mail was sent"

**Comment by jdubois** [comment]
Merged with https://github.com/jhipster/generator-jhipster/commit/dbe65a606f37d6311d2b64e0fd78e5141b18d56b

**Pull request closed without merging**